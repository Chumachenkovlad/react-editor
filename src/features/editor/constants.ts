
export const EDITABLE_CLASS = 'editable';
