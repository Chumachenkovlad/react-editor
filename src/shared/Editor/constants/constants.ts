export const EDITABLE_CLASS = "editable";
export const TEXTAREA_CONTAINER_CLASS = "EditorTextarea-container";
