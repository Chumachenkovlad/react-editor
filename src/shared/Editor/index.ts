export { ConnectedEditor } from './components/ConnectedEditor'
