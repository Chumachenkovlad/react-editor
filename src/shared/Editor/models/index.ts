export * from "./word";
export * from "./formatting";
export * from "./editable-word";
export * from "./common";
