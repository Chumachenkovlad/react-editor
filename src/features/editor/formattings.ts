import { Formatting, FormattingTypes } from "./models";
import { EditorWidgets } from "./components/widgets/widget.model";

export const FORMATTINGS: Formatting[] = [
  {
    key: "bold",
    icon: "format_bold",
    prop: "style.fontWeight",
    appliedValue: "bold",
    type: FormattingTypes.Toggler,
    description: "Bold"
  },
  {
    key: "italic",
    icon: "format_italic",
    prop: "style.fontStyle",
    appliedValue: "italic",
    type: FormattingTypes.Toggler,
    description: "Italic"
  },
  {
    key: "underlined",
    icon: "format_underlined",
    prop: "style.textDecoration",
    appliedValue: "underline",
    type: FormattingTypes.Toggler,
    description: "Underlined"
  },
  {
    key: "background",
    prop: "style.background",
    icon: "color_lens",
    type: FormattingTypes.Widget,
    widgetType: EditorWidgets.ColorPicker,
    description: "Change Background Color"
  },
  {
    key: "color",
    prop: "style.color",
    icon: "text_format",
    type: FormattingTypes.Widget,
    widgetType: EditorWidgets.ColorPicker,
    description: "Change Font Color"
  },
  {
    key: "synonym",
    icon: "edit",
    prop: "innerText",
    type: FormattingTypes.Widget,
    widgetType: EditorWidgets.SynonymsPicker,
    description: "Pick synonym from list"
  },
  {
    key: "fontSize",
    icon: "format_size",
    prop: "style.fontSize",
    type: FormattingTypes.Widget,
    widgetType: EditorWidgets.Input,
    description: "Change Font Size (10px, 1em, etc)"
  }
].map((formattings, index) => ({ ...formattings, index, value: "" }));

export const isTogglerFormatting = ({ type }: Formatting) =>
  type === FormattingTypes.Toggler;
