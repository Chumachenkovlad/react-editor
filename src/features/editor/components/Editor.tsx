import { Nillable, EditableWord, Formatting } from "../models";
import { FORMATTINGS } from "../formattings";
import { getWordValue } from "../utils";
import { set } from "lodash";
import EditorTextarea from "./EditorTextarea";
import EditorToolbar from "./EditorToolbar";

import React, { ReactElement, useState } from "react";

export interface IEditorState {
  selectedWord: Nillable<EditableWord>;
}

export default function Editor(): ReactElement {
  const [selectedWord, setSelectedWord] = useState<Nillable<EditableWord>>(
    null
  );

  function changeWordFormatting(formatting: Formatting) {
    if (!selectedWord) {
      return;
    }

    const { element, formattings } = selectedWord;
    const { property, style, value } = formatting;

    if (property) {
      element.setAttribute(property, value as string);
    }

    if (style) {
      set(element.style, style, value);
    }

    const prevFormattings = formattings.filter(
      ({ key }) => key !== formatting.key
    );

    setSelectedWord({
      formattings: [...prevFormattings, formatting],
      element
    });
  }

  function selectWord(element: HTMLElement) {
    const formattings = FORMATTINGS.map(formatting => {
      const value = getWordValue(element, formatting);
      return { ...formatting, value };
    });

    setSelectedWord({ element, formattings });
  }

  return (
    <div>
      <EditorToolbar
        changeWordFormatting={changeWordFormatting}
        selectedWord={selectedWord}
      />
      <EditorTextarea selectWord={selectWord} />
    </div>
  );
}
