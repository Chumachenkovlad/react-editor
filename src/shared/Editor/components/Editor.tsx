import { Nillable, EditableWord, Formatting } from "../models";
import { get } from "lodash";
import EditorTextarea from "./EditorTextarea";
import EditorToolbar from "./EditorToolbar";

import React, { ReactElement, useState, useRef } from "react";
import { createStorage } from "../../api/storage";
import { DEFAULT_CONTENT } from "../constants/default-content";
import { FORMATTINGS } from "../constants/formattings";
import { applyFormatting } from '../helpers/apply-formatting.func';


export interface IEditorState {
  selectedWord: Nillable<EditableWord>;
}

const storage = createStorage("editor-storage", DEFAULT_CONTENT);
const initialHtml = { __html: storage.getItem() };

export default function Editor(): ReactElement {
  const textareaRef = useRef<HTMLDivElement>(null);
  const [selectedWord, setSelectedWord] = useState<Nillable<EditableWord>>(
    null
  );

  function changeWordFormatting(formatting: Formatting) {
    if (!selectedWord) {
      return;
    }

    const { element, formattings } = selectedWord;

    applyFormatting(element, formatting);

    const prevFormattings = formattings.filter(
      ({ key }) => key !== formatting.key
    );

    setSelectedWord({
      formattings: [...prevFormattings, formatting],
      element
    });

    if (textareaRef.current) {
      storage.setItem(textareaRef.current.innerHTML);
    }
  }

  function selectWord(element: HTMLElement) {
    const formattings = FORMATTINGS.map(formatting => {
      const value = get(element, formatting.prop);
      return { ...formatting, value };
    });

    setSelectedWord({ element, formattings });
  }

  return (
    <div className="Editor">
      <EditorToolbar
        changeWordFormatting={changeWordFormatting}
        selectedWord={selectedWord}
      />
      <div>
        <EditorTextarea selectWord={selectWord}>
          <div
            className="EditorTextarea-container"
            ref={textareaRef}
            dangerouslySetInnerHTML={initialHtml}
          ></div>
        </EditorTextarea>
      </div>
    </div>
  );
}
