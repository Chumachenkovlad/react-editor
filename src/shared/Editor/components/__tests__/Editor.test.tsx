

import React from "react";
import { EDITABLE_CLASS } from 'shared/Editor/constants/constants';
import { RenderResult, render, fireEvent } from '@testing-library/react';
import Editor, { IEditorState } from '../Editor';
import { FORMATTINGS } from "../../constants/formattings";

const getElement = () => {
  const el = document.createElement('span');
  el.classList.add(EDITABLE_CLASS);
  return el
}

describe('<Editor>', () => {
  const selectedWord = {
    element: getElement(),
    formattings: FORMATTINGS
  }
  const htmlContent = 'some text';
  const updateContentSpy = jest.fn();
  const selectWordSpy = jest.fn();
  const updateFormattingSpy = jest.fn();
  let renderResult: RenderResult;

  const props: IEditorState = {
    htmlContent,
    selectedWord: selectedWord,
    updateContent: updateContentSpy,
    selectWord: selectWordSpy,
    updateFormatting: updateFormattingSpy
  }

  beforeEach(() => {
    renderResult = render(<Editor {...props} />)
  })

  test('should be rendered', () => {
    expect(renderResult.container).toMatchSnapshot();
  })

  test('click on toggler button should updateFormatting', () => {
    const index = 0;
    const formatting = FORMATTINGS[index];
    const togglerBtn = renderResult.container.querySelectorAll('button')[index];
    fireEvent.click(togglerBtn);
    expect(updateFormattingSpy).toHaveBeenCalledWith({
      ...formatting,
      value: formatting.appliedValue
    });

    expect(updateContentSpy).toHaveBeenCalledWith(htmlContent);
  })

})
