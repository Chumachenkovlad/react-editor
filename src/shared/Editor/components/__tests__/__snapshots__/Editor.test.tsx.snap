// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor> should be rendered 1`] = `
<div>
  <div
    class="Editor"
  >
    <div
      class="EditorToolbar"
    >
      <div
        class="MuiButtonGroup-root"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedText MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          title="Bold"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              format_bold
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedText MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          title="Italic"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              format_italic
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedText MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          title="Underlined"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              format_underlined
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedText MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          title="Change Background Color"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              color_lens
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedText MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          title="Change Font Color"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              text_format
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedText MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          title="Pick synonym from list"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              edit
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedText MuiButtonGroup-groupedTextHorizontal MuiButtonGroup-groupedText MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          title="Change Font Size (10px, 1em, etc)"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              format_size
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div>
      <div
        class="EditorTextarea"
        contenteditable="true"
      >
        <div
          class="EditorTextarea-container"
        >
          some text
        </div>
      </div>
    </div>
  </div>
</div>
`;
